src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+                _source = source;
+                _introducedLocalNames = new List<string>();
+                _cancellationToken = cancellationToken;
+            }
+
+            public bool TryConvert(out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // Do not try refactoring queries with comments of conditional compilation in them.
+                // We can consider supporting queries with comments in the future.
+                if (_source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                        SyntaxKind.SingleLineCommentTrivia,
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    _source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // Bail out if there is no chance to convert it even with a local function.
+                if (!CanTryConvertToLocalFunction() ||
+                    !TryCreateStackFromQueryExpression(out QueryExpressionProcessingInfo queryExpressionProcessingInfo))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if ((TryConvertInternal(queryExpressionProcessingInfo, out documentUpdateInfo) ||
+                    TryReplaceWithLocalFunction(out documentUpdateInfo)) && // second attempt: at least to a local function
+                    !_semanticModel.GetDiagnostics(_source.Span, _cancellationToken).Any())
+                {
+                    return true;
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private StatementSyntax ProcessClause(CSharpSyntaxNode node, StatementSyntax statement, out StatementSyntax extraStatementToAddAbove)
+            {
+                extraStatementToAddAbove = default;
+                switch (node.Kind())
+                {
+                    case SyntaxKind.WhereClause:
+                        return SyntaxFactory.Block(SyntaxFactory.IfStatement(((WhereClauseSyntax)node).Condition.WithAdditionalAnnotations(Simplifier.Annotation).WithoutTrivia(), statement));
+                    case SyntaxKind.FromClause:
+                        var fromClause = (FromClauseSyntax)node;
+                        return SyntaxFactory.ForEachStatement(
+                                    fromClause.Type ?? VarNameIdentifier,
+                                    fromClause.Identifier,
```

Comment:
>fromClause.Identifier, [](start = 36, length = 22)

This expands scope of this identifier that can be unsafe.

Reply:
We make those checks when parsing the query expression using ContainsIdentifier.
1. We do not allow duplicate identifiers.
2. If converting in for the foreach(var x in query),
we check that there are no identifiers are in the scope of the query statement.

Do you see other cases to handle?

---
In reply to: [178361346](https://github.com/dotnet/roslyn/pull/25362#discussion_r178361346) [](ancestors = 178361346)

Reply:
> Do you see other cases to handle?

It looks like the callers check for this situation.

---
In reply to: [178366178](https://github.com/dotnet/roslyn/pull/25362#discussion_r178366178) [](ancestors = 178366178,178361346)